packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -458,6 +533,44 @@ rocksdb::Status ArbCore::reorgToMessageCountOrBefore(
         std::get<std::unique_ptr<MachineThread>>(std::move(setup));
     auto& output = new_machine->machine_state.output;
 
+    while (new_machine->machine_state.output.arb_gas_used <
+           selected_machine_output.arb_gas_used) {
+        // Need to run machine until caught up with current
+        // checkpoint
+        MachineExecutionConfig execConfig;
+        execConfig.stop_on_sideload = use_latest;
+
+        // Add messages and run machine
+        auto success = runMachineWithMessages(
+            execConfig, coreConfig.message_process_count, new_machine);
+        if (!success) {
+            std::cerr << "runMachineWithMessages failed" << core_error_string
+                      << "\n";
+            return rocksdb::Status::Aborted();
+        }
+
+        if (core_machine->status() == MachineThread::MACHINE_ERROR) {
+            core_error_string = core_machine->getErrorString();
+            std::cerr << "AVM machine stopped with error: " << core_error_string
+                      << "\n";
+            return rocksdb::Status::Aborted();
+        }
+
+        while (new_machine->nextAssertion().sideload_block_number) {
+            timed_sideload_cache.add(std::make_unique<Machine>(*core_machine));
+
+            // Machine was stopped to save sideload,
+            // start machine back up where it stopped
+            auto machine_success = core_machine->continueRunningMachine();
+            if (!machine_success) {
+                core_error_string = "Error continuing machine thread";
+                machine_error = true;
+                std::cerr << "Error catching up: " << core_error_string << "\n";
+                return rocksdb::Status::Aborted();
+            }
+        }
+    }
+
```

Comment:
I'd add a check that `new_machine.output == selected_output || selected_output.arb_gas_used == 0`.

Reply:
:+1: